File: mobile/src/types/testing.ts
export type TestType = "diagnostic" | "adaptive" | "final";

export type TestQuestionOption = {
  id: number;
  text: string;
};

export type TestQuestion = {
  id: number;
  text: string;
  topic: string;
  difficulty: number;
  options: TestQuestionOption[];
};

export type StartDiagnosticResponse = {
  attempt_id: number;
  test_id: number;
  test_title: string;
  test_type: TestType;
  max_questions: number;
};

export type NextQuestionResponse = {
  attempt_id: number;
  test_type: TestType;
  answered_questions: number;
  max_questions: number;
  is_finished: boolean;
  current_theta: number;
  question: TestQuestion | null;
};

export type SubmitAnswerPayload = {
  question_id: number;
  selected_option_id: number;
};

export type SubmitAnswerResponse = {
  attempt_id: number;
  test_type: TestType;
  is_correct: boolean;
  answered_questions: number;
  max_questions: number;
  is_finished: boolean;
  theta_before: number;
  theta_after: number;
};

export type ErrorProfileItem = {
  topic: string;
  total_questions: number;
  wrong_answers: number;
  accuracy_percent: number;
};

export type RecommendationCardItem = {
  category: string;
  reason: string;
  suggested_activity: string;
  estimated_time: string;
  priority: "High" | "Medium" | "Low" | string;
  source?: string;
};

export type AttemptResultResponse = {
  attempt_id: number;
  test_type: TestType;
  started_at: string;
  finished_at: string;
  score_percent: number;
  level_result: string;
  cefr: string;
  theta_final: number;
  total_questions: number;
  correct_answers: number;
  previous_score_percent: number | null;
  score_delta: number | null;
  previous_theta_final: number | null;
  theta_delta: number | null;
  insight: string;
  weak_topics: string[];
  recommendations: string[];
  recommendation_cards?: RecommendationCardItem[];
  error_profile?: ErrorProfileItem[];
};

File: backend/app/schemas/testing.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.db.models import TestType


class TestAttemptStartResponse(BaseModel):
    attempt_id: int
    test_id: int
    test_title: str
    test_type: TestType
    max_questions: int


class QuestionOptionPublic(BaseModel):
    id: int
    text: str

    model_config = ConfigDict(from_attributes=True)


class NextQuestionPayload(BaseModel):
    id: int
    text: str
    topic: str
    difficulty: int
    options: list[QuestionOptionPublic]


class NextQuestionResponse(BaseModel):
    attempt_id: int
    test_type: TestType
    answered_questions: int
    max_questions: int
    is_finished: bool
    current_theta: float
    question: NextQuestionPayload | None = None


class SubmitAnswerRequest(BaseModel):
    question_id: int = Field(gt=0)
    selected_option_id: int = Field(gt=0)


class SubmitAnswerResponse(BaseModel):
    attempt_id: int
    test_type: TestType
    is_correct: bool
    answered_questions: int
    max_questions: int
    is_finished: bool
    theta_before: float
    theta_after: float


class ErrorProfileItem(BaseModel):
    topic: str
    total_questions: int
    wrong_answers: int
    accuracy_percent: float


class AttemptResultResponse(BaseModel):
    attempt_id: int
    started_at: datetime
    finished_at: datetime
    score_percent: float
    level_result: str
    theta_final: float
    total_questions: int
    correct_answers: int
    insight: str
    weak_topics: list[str]
    recommendations: list[str]
    error_profile: list[ErrorProfileItem] = Field(default_factory=list)

